toolchain/check/context.h

```diff
@@ -329,9 +343,13 @@ class Context {
     SemIR::NameScopeId scope_id;
 
     // Names which are registered with name_lookup_, and will need to be
-    // deregistered when the scope ends.
+    // unregistered when the scope ends.
     llvm::DenseSet<IdentifierId> names;
 
+    // Whether a `returned var` was introduced in this scope, and needs to be
+    // unregistered when the scope ends.
+    bool has_returned_var = false;
```

Comment:
Why is this tracked separately from `return_scope_stack_.back().returned_var`? That is, it looks like checking there would give the same result; what prevented its use?

Reply:
This is checking whether there's a `returned var` in the current scope, whereas the `return_scope_stack` value is checking whether there's a `returned var` for the current function. The difference is important because a function can have multiple `returned var`s in different scopes, and we regain the ability to declare a `returned var` when we leave the scope that the current `returned var` is declared within.

See [this section of the design](https://github.com/carbon-language/carbon-lang/blob/trunk/proposals/p0257.md#:~:text=We%20propose%20a%20set%20of%20restrictions%20to%20give%20simple%20and%20understandable%20behavior%20which%20remains%20reasonably%20expressive.) for the rules this is implementing, and `check/testdata/returned_var_scope.carbon` and `check/testdata/fail_returned_var_shadow.carbon`.